nm dispatch: Fix race when modifying dispatch script

cathay4t opened the pull request:
Previously we reapply/activate the NmConnection before modifying the
dispatcher script, this would cause NetworkManager invoked old dispatcher
script content.

Fixed by modify the dispatcher script before NmConnection actions, also
invoke `File::sync_all()` to ensure dispatch script been sync to file
system for NetworkManager-dispatcher.service to read afterwards.

Integration test case included.

Resolves: https://issues.redhat.com/browse/RHEL-94840

cathay4t committed:
Run cargo clippy and cargo fmt

Signed-off-by: Gris Ge <fge@redhat.com>

bengal commented on rust/src/lib/nm/query_apply/dispatch.rs:
```diff
@@ -115,6 +115,7 @@ fn write_execute_file(file_path: &str, content: &str) -> std::io::Result<()> {
         .mode(0o744)
         .open(file_path)?;
     fd.write_all(content.as_bytes())?;
+    fd.sync_all()?;
```
Why is it needed to ensure the file is written to disk? That is normally not necessary for inter-process synchronization: you can just write to the file, close it (or flush it) and then other processes can see the changes. Waiting for sync_all() adds an unnecessary and potentially long delay.

cathay4t replied:
The close(2) manpage has:

```
A successful close does not guarantee that the data has been successfully saved
to disk, as the kernel uses the buffer cache to defer writes.  Typically,
filesystems do not flush buffers when a file is closed.  If you need to be sure
that the data is physically stored on the underlying disk, use fsync(2).  (It
will depend on the disk hardware at this point.)
```

Yes, you are right, `close()` only make sure it reach filesystem buffer which is enough for other process to see the change.

Removed.

cathay4t committed:
nm dispatch: Fix race when modifying dispatch script

Previously we reapply/activate the NmConnection before modifying the
dispatcher script, this would cause NetworkManager invoked old
dispatcher script content.

Fixed by modify the dispatcher script before NmConnection actions.

Integration test case included.

Resolves: https://issues.redhat.com/browse/RHEL-94840

Signed-off-by: Gris Ge <fge@redhat.com>

bengal commented:
LGTM
